cpp/inc/bond/core/maybe.h

```diff
@@ -14,143 +18,249 @@ namespace bond
+    : public detail::maybe_common<T>,
+      private detail::allocator_type<T>::type
+{
+public:
+    using allocator_type = typename detail::allocator_type<T>::type;
+
+    maybe() = default;
+    maybe(const maybe&) = default;
+
+// TODO: specialize std::uses_allocator with conditional on has_allocator && then is_convertable
+
+    maybe(maybe&& that) BOND_NOEXCEPT_IF(std::is_nothrow_move_constructible<typename detail::maybe_common<T>>::value
+                                         && std::is_nothrow_move_constructible<allocator_type>::value)
+        : detail::maybe_common<T>(static_cast<detail::maybe_common<T>&&>(that)),
+          allocator_type(static_cast<allocator_type&&>(that))
+    { }
+
+    /// @brief Move constructor
+    maybe(
+        maybe&& that,
+        const allocator_type& alloc)
+        BOND_NOEXCEPT_IF(std::is_nothrow_move_constructible<typename detail::maybe_common<T>>::value
+                         && std::is_nothrow_copy_constructible<allocator_type>::value)
+        : detail::maybe_common<T>(static_cast<detail::maybe_common<T>&&>(that)),
+          allocator_type(alloc)
+    { }
+
+    explicit maybe(const T& value)
+        : detail::maybe_common<T>(value),
+          allocator_type(allocator_type { })
+    { }
+
+    explicit maybe(const allocator_type& alloc)
+        : detail::maybe_common<T>(),
+          allocator_type(alloc)
+    { }
+
+    maybe(const T& value, const allocator_type& alloc)
+        : detail::maybe_common<T>(value, alloc),
+          allocator_type(alloc)
+    { }
+
+    maybe& operator=(const maybe&) = default;
 
     /// @brief Assign a value T
     maybe& operator=(const T& value)
     {
-        _value = value;
-        _nothing = false;
+        // TODO: yes/no on passing allocator?
+        this->_value.emplace(value, static_cast<const allocator_type&>(*this));
         return *this;
     }
 
+    /// @brief If this object contains nothing, set it to a default instance
+    /// of T; otherwise, preserve the existing value. Then, return a
+    /// reference to the value.
+    T& set_value()
+    {
+        if (this->is_nothing())
+        {
+            this->_value.emplace(static_cast<const allocator_type&>(*this));
+        }
 
-    /// @brief Convert to constant reference to value
-    /// @throw NothingException if the object contains nothing
-    operator const T&() const
+        return this->_value.get();
+    }
+
+    /// @brief Swap this object with that object
+    void swap(maybe& that)
     {
-        return value();
+        using std::swap;
+        swap(this->_value, that._value);
+        swap(static_cast<allocator_type&>(*this), static_cast<allocator_type&>(that));
     }
 
-private:
-    T       _value;
-    bool    _nothing;
-};
+    /// @brief Compare for equality
+    bool operator==(const maybe& that) const
+    {
+        return this->_value == that._value
+            &&
+        static_cast<const allocator_type&>(*this)
```

Comment:
Do we really care about allocator equality?

Reply:
I'm not sure. Looking at MSVC and libstdc++'s vector and map, they don't care, so I think we can drop the allocator comparison too.

Reply:
Yeah, the `operator==` in this case is about value equality.